### PR TITLE
Improvements to Felix liveness/readiness reporting

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

* Tune the liveness/readiness timeouts to avoid false positives.
* Reset liveness from within calc graph's tight loops.
* Add capability to override any problematic timeout.
  * Refactor configuration handling to make it easier to handle changes without restart.
* Make names of health probes uniform `CamelCase`

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Felix now supports overriding the timeouts of its internal readiness/liveness watchdog.  This is useful for dealing with issues "in prod" without needing a new release.  The timeouts have also been tuned to reduce false positives.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: typha/pkg/snapcache/cache.go
Comment (song-jiang) [inline typha/pkg/snapcache/cache.go:156]: healthInterval hasn't been replace here.

File: api/pkg/apis/projectcalico/v3/felixconfig.go
Comment (caseydavenport) [inline api/pkg/apis/projectcalico/v3/felixconfig.go:258]: How feasible is it to make this a structured field?

For example:

```yaml
- name: InternalDataplaneMainLoop
  value: 30s
```
or

```
timeoutOverrides:
  internalDataplaneMainLoop: 30s
```

Or similar? I think these comma-separated string lists are fairly awkward to use. Their main benefit IIRC is that they are easier to encode into an env var, but we're pushing away from that being the primary means of configuring Felix.

I understand this might be a challenge given Felix's config merging machinery, but figured I'd ask!

Comment (song-jiang) [inline api/pkg/apis/projectcalico/v3/felixconfig.go:258]: Shall we make the the field a type which contains timeout values for all active components? 
```
type HealthTimeoutOverrides struct {
    CalculationGraph *metav1.Duration
    ....
```

The benefit is that the API is much clearer to user, and it you get the default FelixConfiguration, it will show you all the default values for timeout.

**Review by song-jiang — COMMENTED** [review]
Looking good. Just one question on if we should put timeout settings into an API structure and a few nits.

**Review by caseydavenport — COMMENTED** [review]
Some minor questions - I haven't gotten through all of it but I think @song-jiang is taking a closer look.

File: felix/calc/async_calc_graph.go
Comment (fasaxc) [reply felix/calc/async_calc_graph.go:94]: I didn't rename it, I added a new constant for the timeout.  That was due to a change in an earlier revision; I could back it out but maybe it's clearer this way anyway

Comment (song-jiang) [reply felix/calc/async_calc_graph.go:94]: I see. My mistake.

File: api/pkg/apis/projectcalico/v3/felixconfig.go
Comment (fasaxc) [reply api/pkg/apis/projectcalico/v3/felixconfig.go:258]: I should be able to make it a list or a map but I'm reluctant to pre-define all the health report names since they're currently allocated in a dispersed way.  I can't easily write a UT to ensure that none are missed either since the ones that are active depend on build and run-time configuration.  

I envisioned this as a safety valve that we'd tell a user to set in an emergency rather than something that users would necessarily discover on their own.

Comment (fasaxc) [reply api/pkg/apis/projectcalico/v3/felixconfig.go:258]: (The main reason I didn't do it was naming convention mismatch with k8s; seemed a bit odd to use `lowerCamel` for health report names. Could live with UpperCamel, as Casey suggested.)

Comment (caseydavenport) [reply api/pkg/apis/projectcalico/v3/felixconfig.go:258]: > pre-define all the health report names since they're currently allocated in a dispersed way

I think I'm begrudgingly OK with the argument that this is an escape valve that is meant more for developers to help customers than it is for users to set themselves. The problem is that we mix those sorts of fields with proper user-facing fields in the same resource so it's not super obvious which are which.

File: libcalico-go/lib/validator/v3/validator.go
Comment (fasaxc) [reply libcalico-go/lib/validator/v3/validator.go]: Removed this but added UT for the new validation

File: api/pkg/apis/projectcalico/v3/felixconfig.go
Comment (fasaxc) [reply api/pkg/apis/projectcalico/v3/felixconfig.go]: Done, made it a slice

Comment (fasaxc) [reply api/pkg/apis/projectcalico/v3/felixconfig.go]: I made this a slice of override objects and improved the documentation

**Review by caseydavenport — APPROVED** [review]

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Pull request merged**
